### ISSUE TITLE
TransferManager::DownloadFile: If filename includes a non-existent directory does not produce any error

### ISSUE THREAD

**Issue by svagionitis** [aws/aws-sdk-cpp#578]
Hi all,

When I use the `TransferManager::DownloadFile` to download a file, and the filename includes a non-existent directory, I don't get any error back that there was some problem. The download continues with no error and I suppose is downloaded somewhere in the memory.

The comment on the header file for `TransferManager::DownloadFile` does not specify if the `writeToFile` is just a filename or a pathname.

```
           /**                                                                  
             * Downloads the contents of bucketName/keyName in S3 to the file specified by writeToFile. This will perform a GetObject operation.
             */                                                                  
            std::shared_ptr<TransferHandle> DownloadFile(const Aws::String& bucketName, 
                                                         const Aws::String& keyName, 
                                                         const Aws::String& writeToFile, 
                                                         const DownloadConfiguration& downloadConfig = DownloadConfiguration());
```

A temporary solution that I found for my case is to create the directory before I call the `TransferManager::DownloadFile`, like the following

```
Aws::FileSystem::CreateDirectoryIfNotExists(Dirname (writeToFile));
TransferManager::DownloadFile(bucketName, keyName, writeToFile);
```

I was wondering if this behavior is intended like this or it's a bug?

Regards,

Stavros

**Comment by marcomagdy** [comment]
Sounds like a good candidate for a pull request :)

**Comment by diablodale** [comment]
Adding more details: with SDK v1.1.49, downloading a large file with `DownloadFile()` to a path containing a folder that doesn't yet exist causes an exception crash. Small files do not crash, instead they report IN_PROGRESS and COMPLETE yet produce no resulting file on the hard drive. I suspect this is due to the large file using the multipart download codepath.

The exception is in `aws-sdk-cpp\aws-cpp-sdk-transfer\source\transfer\TransferHandle.cpp` line 296:
```cpp
m_downloadStream->good()
```

Please note that this bug makes the behavior of `DownloadFile()` and `DownloadToDirectory()` different. The latter creates any needed destination directories.

The preferred workaround won't work. You can not create the directories in the IN_PROGRESS state of `transferStatusUpdatedCallback` due to https://github.com/aws/aws-sdk-cpp/issues/653.

Therefore, one must put the create of directory(s) immediately before `DownloadFile()`. Naturally, the sideaffect of this fallback workaround is the creation of directories regardless of the download starting, progressing, failing, completing, etc.

**Comment by singku** [comment]
This should have been fixed with this commit: 8082ace481931523a8ed8967f1cda163443dcee9
